packages/engine-core/src/library/ExitHooks.ts

```diff
@@ -48,20 +49,34 @@ export class ExitHooks {
     return this.idToListenerMap.get(id)
   }
 
-  private installHook(event: string, shouldExit = false) {
-    process.once(event, async (code) => {
-      debug(`exit event received: ${event}`)
-      for (const listener of this.idToListenerMap.values()) {
-        await listener()
-      }
+  private installExitEventHook(event: 'beforeExit' | 'exit') {
+    // Note: TypeScript isn't able to narrow type arguments on overloaded functions.
+    process.once(event as any, this.exitLikeHook)
+  }
 
-      this.idToListenerMap.clear()
+  private installExitSignalHook(signal: NodeJS.Signals) {
+    process.once(signal, async (signal) => {
+      await this.exitLikeHook(signal)
+      const isSomeoneStillListening = process.listenerCount(signal) > 0
 
-      // only exit, if only we are listening
-      // if there is another listener, that other listener is responsible
-      if (shouldExit && process.listenerCount(event) === 0) {
-        process.exit(code)
+      // Only exit when there are no listeners left for this signal.
+      // If there is another listener, that other listener is responsible for exiting.
+      if (isSomeoneStillListening) {
+        return
       }
+
+      // the usual way to exit with a signal is to add 128 to the signal number
+      const exitCode = os.constants.signals[signal] + 128
```

Comment:
People might rely on the old exitCode, not sure if we want to do this change in a standard release?

Reply:
Considering this is a direct consequence of Node.js changing its behavior, that's not up to us.
Afaik, as long as we don't turn a `0` exit code into another number, everything's fine

Reply:
I am with Alberto on this, right now, you applications exit code changes to 0 if you set our `beforeExit` hook, this change would restore them to what they are by default.

Reply:
Not sure about this since it was always 0 until this PR.
Also, we would at least need a test for this? It's an important bit that is still untested based on the CI run here?

Reply:
It was not always 0. We set our signal handlers only when you add `beforeExit` event listener. If you don't do it, exit code will be default one, 128 + signal.

Reply:
Which means it was always 0 for people using the `beforeExit` event listener 😄